fix: 193 rate limit issue with pullpush components

alvarosabu opened the pull request:
This PR introduces a partial refactor of the codebase replacing the `customFetch` method with a new local management API client (MAPI) with custom retry logic on all `component` related actions.

Closes #193

alvarosabu committed:
feat: wip mapi client with retry mechanism

alvarosabu committed:
feat: freeze mechanism for retry

alvarosabu committed:
feat: wip concurrency limit

alvarosabu committed:
refactor: replaced custom fetch with mapi client with retry mechanism for component internal tags

alvarosabu committed:
refactor: replace customFetch with mapiClient on all component related operations

alvarosabu committed:
Merge branch 'next' into bugfix/193-rate-limit-issue-with-pullpush-components

alvarosabu committed:
refactor: simplify function signatures by removing unnecessary parameters in component handling functions

- Removed `password` and `region` parameters from `handleTags`, `handleComponentGroups`, `handleWhitelists`, and `upsertComponentPreset` functions, streamlining the function calls and improving clarity.
- Updated corresponding tests to reflect the changes in function signatures, ensuring consistency across the codebase.

alvarosabu committed:
refactor(tests): remove unused mock variables in operations tests

- Eliminated `mockPassword` and `mockRegion` constants from `operations.test.ts` as they are no longer necessary, streamlining the test setup.
- Updated `operations.ts` to remove the import of `RegionCode`, reflecting the recent changes in function signatures and improving code clarity.

imranolas commented on package.json:
```diff
@@ -43,6 +43,7 @@
     "dotenv": "^16.5.0",
     "json-schema-to-typescript": "^15.0.4",
     "ohash": "^2.0.11",
+    "p-limit": "^6.2.0",
```
Is this used?

imranolas commented on src/api.ts:
```diff
@@ -1,51 +1,189 @@
+      if (options?.verbose) {
+        console.log(`⏳ ${path} - Waiting for rate limit to be resolved`);
+      }
+      await new Promise<void>((resolve) => {
+        const checkFreeze = setInterval(() => {
+          if (!state.freeze) {
+            clearInterval(checkFreeze);
+            resolve();
+          }
+        }, 50);
+      });
```
A nicer alternative to this async loop would be to hold the value of state.freeze as a promise which can be awaited. This would avoid having to loop to check the state.

There's also a handy API for this. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers

alvarosabu replied:
Lets tackle this in an upcoming PR. @imranolas

alvarosabu committed:
chore: remove p-limit dependency from package.json and pnpm-lock.yaml

- Eliminated `p-limit` from `package.json` and corresponding entries in `pnpm-lock.yaml` to streamline dependencies and reduce package size.
